File: src/cn/yiiguxing/plugin/translate/action/TranslateAction.java
package cn.yiiguxing.plugin.translate.action;

import cn.yiiguxing.plugin.translate.TranslationUiManager;
import cn.yiiguxing.plugin.translate.Utils;
import cn.yiiguxing.plugin.translate.ui.Icons;
import cn.yiiguxing.plugin.translate.ui.TranslationBalloon;
import com.intellij.codeInsight.highlighting.HighlightManager;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.RangeMarker;
import com.intellij.openapi.editor.markup.EffectType;
import com.intellij.openapi.editor.markup.RangeHighlighter;
import com.intellij.openapi.editor.markup.TextAttributes;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Disposer;
import com.intellij.openapi.util.TextRange;
import com.intellij.ui.JBColor;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.util.ArrayList;

public class TranslateAction extends AutoSelectAction implements DumbAware {

    private static final TextAttributes HIGHLIGHT_ATTRIBUTES;

    static {
        TextAttributes attributes = new TextAttributes();
        attributes.setBackgroundColor(new JBColor(new Color(0xFFE4E4FF), new Color(0xFF344134)));
        attributes.setEffectType(EffectType.LINE_UNDERSCORE);
        attributes.setEffectColor(new JBColor(0xFFEE6000, 0xFFCC7832));

        HIGHLIGHT_ATTRIBUTES = attributes;
    }

    /**
     * @param checkSelection 指定是否检查手动选择的文本。<code>true</code> - 如果有手动选择文本，
     *                       则忽略<code>autoSelectionMode</code>, <code>false</code> - 将忽略手动选择的文本。
     */
    public TranslateAction(boolean checkSelection) {
        super(Icons.Translate, checkSelection);
    }

    /**
     * 自动从最大范围内取词，忽略选择
     */
    public TranslateAction() {
        this(false);
    }

    /**
     * 返回取词模式
     */
    @NotNull
    @Override
    protected AutoSelectionMode getAutoSelectionMode() {
        return AutoSelectionMode.INCLUSIVE;
    }

    @Override
    protected void onUpdate(AnActionEvent e, boolean active) {
        e.getPresentation().setEnabledAndVisible(active);
    }

    @Override
    protected void onActionPerformed(AnActionEvent e, @NotNull final Editor editor, @NotNull TextRange selectionRange) {
        final String queryText = Utils.splitWord(editor.getDocument().getText(selectionRange));
        if (!Utils.isEmptyOrBlankString(queryText)) {
            final Project project = editor.getProject();
            final ArrayList<RangeHighlighter> highlighters = new ArrayList<RangeHighlighter>();
            final HighlightManager highlightManager = project == null ? null : HighlightManager.getInstance(project);

            if (highlightManager != null) {
                highlightManager.addRangeHighlight(editor, selectionRange.getStartOffset(),
                        selectionRange.getEndOffset(), HIGHLIGHT_ATTRIBUTES, true, highlighters);
            }

            RangeMarker caretRangeMarker = createCaretRangeMarker(editor, selectionRange);
            TranslationBalloon translationBalloon = TranslationUiManager.getInstance()
                    .showTranslationBalloon(editor, caretRangeMarker, queryText);

            if (!highlighters.isEmpty() && highlightManager != null) {
                Disposer.register(translationBalloon.getDisposable(), new Disposable() {
                    @Override
                    public void dispose() {
                        for (RangeHighlighter highlighter : highlighters) {
                            highlighter.dispose();
                        }
                    }
                });
            }
        }
    }

    @NotNull
    private RangeMarker createCaretRangeMarker(@NotNull Editor editor, @NotNull TextRange selectionRange) {
        RangeMarker myCaretRangeMarker = editor.getDocument().createRangeMarker(Utils.requireNonNull(selectionRange));
        myCaretRangeMarker.setGreedyToLeft(true);
        myCaretRangeMarker.setGreedyToRight(true);

        return myCaretRangeMarker;
    }

}

File: src/main/kotlin/cn/yiiguxing/plugin/translate/action/TranslateAction.kt
package cn.yiiguxing.plugin.translate.action

import cn.yiiguxing.plugin.translate.service.TranslationUIManager
import cn.yiiguxing.plugin.translate.ui.BalloonPositionTracker
import cn.yiiguxing.plugin.translate.util.SelectionMode
import cn.yiiguxing.plugin.translate.util.createCaretRangeMarker
import cn.yiiguxing.plugin.translate.util.processBeforeTranslate
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.actionSystem.PopupAction
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.ScrollType
import com.intellij.openapi.editor.markup.*
import com.intellij.openapi.project.DumbAware
import com.intellij.openapi.util.Disposer
import com.intellij.openapi.util.TextRange
import com.intellij.ui.JBColor

/**
 * 翻译动作
 *
 * @param checkSelection 指定是否检查手动选择的文本。`true` - 如果有手动选择文本, 则忽略`autoSelectionMode`, `false` - 将忽略手动选择的文本。
 */
open class TranslateAction(checkSelection: Boolean = false) :
    AutoSelectAction(checkSelection),
    PopupAction,
    DumbAware {

    override val selectionMode
        get() = SelectionMode.INCLUSIVE

    override fun onActionPerformed(event: AnActionEvent, editor: Editor, selectionRange: TextRange) {
        val project = editor.project ?: CommonDataKeys.PROJECT.getData(event.dataContext)
        val markupModel = editor.markupModel
        val selectionModel = editor.selectionModel
        val isColumnSelectionMode = editor.caretModel.caretCount > 1

        val text: String
        val starts: IntArray
        val ends: IntArray
        if (selectionModel.hasSelection(true) && isColumnSelectionMode) {
            starts = selectionModel.blockSelectionStarts
            ends = selectionModel.blockSelectionEnds
            text = selectionModel.getSelectedText(true)?.processBeforeTranslate() ?: return
        } else {
            starts = intArrayOf(selectionRange.startOffset)
            ends = intArrayOf(selectionRange.endOffset)
            text = editor.document.getText(selectionRange).processBeforeTranslate() ?: return
        }

        //this logic is also used in ShowTranslationDialogAction
        val currentNewTD = TranslationUIManager.instance(project).currentTranslationDialog()
        if (currentNewTD != null) {
            currentNewTD.translate(text)
            return
        }

        val startLine by lazy { editor.offsetToVisualPosition(selectionRange.startOffset).line }
        val endLine by lazy { editor.offsetToVisualPosition(selectionRange.endOffset).line }
        val highlightAttributes = if (starts.size > 1 || startLine == endLine) {
            HIGHLIGHT_ATTRIBUTES
        } else {
            MULTILINE_HIGHLIGHT_ATTRIBUTES
        }

        val highlighters = ArrayList<RangeHighlighter>(starts.size)
        try {
            for (i in starts.indices) {
                highlighters += markupModel.addRangeHighlighter(
                    starts[i],
                    ends[i],
                    HighlighterLayer.SELECTION - 1,
                    highlightAttributes,
                    HighlighterTargetArea.EXACT_RANGE
                )
            }

            editor.scrollingModel.scrollToCaret(ScrollType.MAKE_VISIBLE)

            val caretRangeMarker = editor.createCaretRangeMarker(selectionRange)
            val tracker = BalloonPositionTracker(editor, caretRangeMarker)
            val balloon = TranslationUIManager.showBalloon(editor, text, tracker)

            if (balloon.disposed) {
                markupModel.removeHighlighters(highlighters)
            } else {
                Disposer.register(balloon) { markupModel.removeHighlighters(highlighters) }
            }
        } catch (thr: Throwable) {
            markupModel.removeHighlighters(highlighters)
        }
    }

    private companion object {
        val EFFECT_COLOR = JBColor(0xFFEE6000.toInt(), 0xFFCC7832.toInt())

        val HIGHLIGHT_ATTRIBUTES: TextAttributes = TextAttributes().apply {
            effectType = EffectType.LINE_UNDERSCORE
            effectColor = EFFECT_COLOR
        }

        val MULTILINE_HIGHLIGHT_ATTRIBUTES: TextAttributes = TextAttributes().apply {
            effectType = EffectType.BOXED
            effectColor = EFFECT_COLOR
        }

        private fun MarkupModel.removeHighlighters(highlighters: Collection<RangeHighlighter>) {
            for (highlighter in highlighters) {
                removeHighlighter(highlighter)
                highlighter.dispose()
            }
        }
    }
}
